vms/evm/predicate/results_test.go

```diff
@@ -4,227 +4,413 @@
-						common.Address{2}: set.NewBits(1, 2, 3),
-					},
+			name: "multiple txs multiple results",
+			input: BlockResults{
+				{1}: {
+					{2}: set.NewBits(1, 2, 3),
+					{3}: set.NewBits(3, 2, 1),
+				},
+				{2}: {
+					{2}: set.NewBits(1, 2, 3),
+					{3}: set.NewBits(3, 2, 1),
 				},
-				{tx: common.Hash{1}, results: PrecompileResults{}},
 			},
-			gets: []getArgs{
-				{tx: common.Hash{1}, addr: common.Address{2}, wantBits: set.Bits{}, wantOK: false},
+			want: []byte{
+				// codecID
+				0x00, 0x00,
+				// BlockResults length
+				0x00, 0x00, 0x00, 0x02,
+				// txHash
+				0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				// PrecompileResults length
+				0x00, 0x00, 0x00, 0x02,
+				// precompile address
+				0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00,
+				// Length of bitset
+				0x00, 0x00, 0x00, 0x01,
+				// bitset
+				0b00001110,
+				// precompile address
+				0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00,
+				// Length of bitset
+				0x00, 0x00, 0x00, 0x01,
+				// bitset
+				0b00001110,
+				// txHash
+				0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				// PrecompileResults length
+				0x00, 0x00, 0x00, 0x02,
+				// precompile address
+				0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00,
+				// Length of bitset
+				0x00, 0x00, 0x00, 0x01,
+				// bitset
+				0b00001110,
+				// precompile address
+				0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x00, 0x00,
+				// Length of bitset
+				0x00, 0x00, 0x00, 0x01,
+				// bitset
+				0b00001110,
 			},
-			wantLen: 1,
 		},
 	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			require := require.New(t)
 
-	for _, tc := range tests {
-		t.Run(tc.name, func(t *testing.T) {
-			req := require.New(t)
-			var br BlockResults
-			for _, s := range tc.sets {
-				br.Set(s.tx, s.results)
-			}
-			for _, g := range tc.gets {
-				got, ok := br.Get(g.tx, g.addr)
-				req.Equal(g.wantOK, ok)
-				req.Equal(g.wantBits, got)
+			got, err := test.input.Bytes()
+			require.NoError(err)
+			require.Equal(test.want, got)
+
+			input := test.input
+			if input == nil {
+				// nil and empty should be considered equivalent
+				input = make(BlockResults)
 			}
-			req.Len(br.TxResults, tc.wantLen)
+
+			parsed, err := ParseBlockResults(got)
+			require.NoError(err)
+			require.Equal(input, parsed)
```

Comment:
I thought about moving this out to a separate test... But I felt like it was overly repetitive.